Fix width of select elements in settings.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First commit is to fix width of select elements in settings page. The width is set to auto so that they can be wide enough for different languages with minimum width set to 325px.
- Second commit is to fix width of select elements in stream settings page. The width is set to auto so that they can be wide enough for different languages with minimum width set to 325px.
- Third commit is to fix the width of dropdown-list-widget for long options.
 <!-- Issue link, or clear description.-->

https://chat.zulip.org/#narrow/stream/101-design/topic/select.20elements.20in.20settings
<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented:
GIFs for settings page -
![select-settings-english](https://user-images.githubusercontent.com/35494118/221632819-7653561d-6a23-4449-bdd7-f1ef0868d801.gif)
![select-settings-german](https://user-images.githubusercontent.com/35494118/221632852-7f95e6c2-0435-4475-a159-0fe5e7490828.gif)
![select-settings-russian](https://user-images.githubusercontent.com/35494118/221632900-53b74617-77ae-4eb2-aa89-109428b9235a.gif)

sahil839 commented:
GIFs for stream settings page -
![stream-settings-english](https://user-images.githubusercontent.com/35494118/221633011-3b84fe40-1fcb-4b38-a970-73e9f6b4562c.gif)
![stream-settings-german](https://user-images.githubusercontent.com/35494118/221633081-c73d4513-6c7e-47ce-b584-2b4a3b19a3ab.gif)
![stream-settings-russian](https://user-images.githubusercontent.com/35494118/221633127-a7fa5653-063d-4915-86a2-94031fcaa449.gif)

sahil839 commented:
Note that for dropdown-list-widget, I have only fixed the behavior for long options in stream settings page. We can fix others if this looks good.

sahil839 commented:
Can we make the page scrollable horizontally in such case? If we do this then user can see the options completely. We do that currently for stream settings.

alya commented:
Sure, that seems reasonable to me.

sahil839 commented:
Added commit to enable horizontal scrolling.
![settings-horizontal-scroll](https://user-images.githubusercontent.com/35494118/221773237-ad980bcc-be44-4e3a-8f86-2f3fd4ebab98.gif)

sahil839 commented:
Also, if the behavior for dropdown-list-widgets (the one for "Who can unsubscribe others" stream setting) seems fine, I can try doing that change for others as well.

alya commented:
> Also, if the behavior for dropdown-list-widgets (the one for "Who can unsubscribe others" stream setting) seems fine, I can try doing that change for others as well.

I didn't notice any problems with it. Anything in particular I should pay attention to?

The only thing is that it would be great if the down-angles looked the same across all dropdowns, or at least within a single settings page.

sahil839 commented:
The things that have changed in the widget are -
- The position of the dropdown menu iopened is shifted slightly above.
- For long options present in the dropdown. the menu is wide enough to fit the longest option, while the button to open the dropdown can be less wide as its width is to fit the selected option.

> The only thing is that it would be great if the down-angles looked the same across all dropdowns, at the very least within a single settings page.
This should be fixed in #23071. I stopped working on it due to it not working correctly for long options, so should be able to work on it once we finalize this.

alya commented:
OK, cool, those changes feel fine to me.

One subtle thing I noticed while testing again is that I hover over an option with the mouse, pressing "Enter" does not select it. It does select it for other dropdowns, which I think is better.

timabbott commented:
What's the next step for this? I see it's still marked as a draft but don't know what the next action is.

timabbott commented:
I skipped the commits and they look generally reasonable. We should consider moving `.settings_select` to be a top-level class in `app_components.css` rather than one that has a duplicate copy in stream settings.

sahil839 commented:
I am still doing the work to update the width of menus for all dropdown list widgets. Will try to update the PR by tonight.

sahil839 commented on web/styles/app_components.css:
```diff
@@ -842,3 +842,15 @@ div.overlay {
+
+        span {
+            width: fit-content;
+        }
```
Added this here because we do not want these changes for dropdown-list-widget in "Move topic" modal.

timabbott replied:
Hmm, what happens if we have these on all dropdown-list-widgets?

I'm wondering if we can express this idea through some sort of option or configuration.

Probably fine to do in #frontend though -- not a blocker for this PR.

sahil839 replied:
Adding this to all widgets, leads to issue in the "Move topic" menu as we have set the postiion of the menu to be `fixed` for that case for it to be completely visible and that results in unexpected UI with the submit buttons appearing on top of the dropdown menu.

Can try adding it as an config option though.

sahil839 commented:
Updated the PR.
- Added commit to fix scrolling in stream settings in narrow width screens.
- Added commit to make dropdown-list-widget toggle button for `New user announcements`, `New stream announcements` and `Default language for code blocks` to have minimum width of 325px.
- Added commit to change the dropdown menu width for - `New user announcements`, `New stream announcements`, `Default language for code blocks` in organization settings, `Owner` dropdown in "Manage bot" modal and `Who can remove others from stream` stream setting, to fit long options.

![stream-settings-dropdown](https://user-images.githubusercontent.com/35494118/223767932-8f1eacce-4600-4447-a753-85003e7384c5.gif)

![bot-owner-dropdown](https://user-images.githubusercontent.com/35494118/223767991-161fd11d-ffff-4b64-9964-80605513656a.gif)

sahil839 commented:
@alya would be good if you could test it once again if the behavior for dropdown list widgets looks good in both normal and narrow screens.

alya commented:
Thanks! We should prevent the bot owner dropdown from going on two lines:

![Screen Shot 2023-03-08 at 10 34 49 AM](https://user-images.githubusercontent.com/2090066/223804438-f6bcb247-81d3-4b34-85aa-53e76fa71d61.png)

alya commented:
Looks good to me otherwise! The follow-ups sound good as well.

sahil839 commented:
```
We should prevent the bot owner dropdown from going on two lines when the selected option is long:
```
I think this is being fixed in #24487. Need to review that PR too once. Will do it today itself.

alya commented:
@timabbott I think you should take a look now. There are some follow-ups, but this PR may be ready for integration.

timabbott commented:
Merged, thanks @sahil839!
